streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/task/CheckpointProcessor.java

```diff
@@ -52,30 +62,31 @@ public class CheckpointProcessor {
 
   @Autowired private SavePointService savePointService;
 
-  public void process(Long appId, CheckPoints checkPoints) {
-    CheckPoints.Latest latest = checkPoints.getLatest();
-    if (latest == null || latest.getCompleted() == null) {
-      return;
-    }
-    CheckPoints.CheckPoint checkPoint = latest.getCompleted();
+  public void process(Long appId, @Nonnull CheckPoints checkPoints) {
+    checkPoints.getLatestCheckpoint().forEach(checkPoint -> process(appId, checkPoint));
+  }
+
+  private void process(Long appId, @Nonnull CheckPoints.CheckPoint checkPoint) {
     Application application = applicationService.getById(appId);
     CheckPointStatus status = checkPoint.getCheckPointStatus();
 
     if (CheckPointStatus.COMPLETED.equals(status)) {
-      String cacheId = appId + "_" + application.getJobId();
-      Long latestId =
-          checkPointCache.get(
-              cacheId,
-              key -> {
-                SavePoint savePoint = savePointService.getLatest(appId);
-                return Optional.ofNullable(savePoint).map(SavePoint::getChkId).orElse(null);
-              });
-
-      if (latestId == null || latestId < checkPoint.getId()) {
-        saveSavepoint(checkPoint, application);
+      String cacheId = getCacheIdForCheckpoint(appId, application.getJobId());
+      Long latestChkId = getLatestCheckpointedId(appId, cacheId);
+      if (shouldStoreAsSavepoint(checkPoint, latestChkId)) {
```

Comment:
> However, if the savepoint store-logic follow checkpoint, here would be some bug-case. If the triggered savepoint is completed after a completed checkpoint whose chk-id is bigger than the savepoint, the savepoint will be discard based on the current logic.

When it happened, the latestChkId is bigger than the savepoint. So `shouldStoreAsSavepoint(checkPoint, latestChkId)` returns false, right?

Reply:
If in this cased you described, the logic will go to `! isStoredSavepoint`.